### ISSUE TITLE
Simplify installation on Unraid

### ISSUE THREAD

**Issue by spiderben25** [photoprism/photoprism#411]
Hi,
I'm trying to install the Docker image on my Unraid server using Dockerhub.
It's working fine but I'm struggling this permission issues. Photoprism runs as root:root and all the files it creates are given the root:root membership.
Unraid uses nobody:users (UID 99 GID 100) for all files on its shares.
Usually, Docker containers compatible with unraid (like linuxserver.io ones) allow the user to passthrough a PUID and PGID as environment variables in order to set correct permissions.
This issue had been raised here: https://github.com/photoprism/photoprism/issues/263 but it was closed. Unfortunately I can't use docker-compose to pass user setting like it was suggested.
Would it be possible to add these parameters as environment variables?
Thanks!

To make myself more clear (bad English, sorry...) :
Here is an example for a compatible Docker image : 
![Capture](https://user-images.githubusercontent.com/2954955/87658982-2da69800-c75d-11ea-93bb-aef71999a0c5.PNG)
and the explanation on how they do it: https://hub.docker.com/r/linuxserver/sonarr

And here are how Photoprism created files currently appear:
![Capture2](https://user-images.githubusercontent.com/2954955/87659500-f4baf300-c75d-11ea-91ce-75ab3978b198.PNG)

**Comment by spiderben25** [comment]
Your solution worked perfectly. Thank you very much, Photoprism is really
awesome!

Le ven. 17 juil. 2020 à 10:50, bobobo1618 <notifications@github.com> a
écrit :

**Comment by ich777** [comment]
This doesn't work very will since you are limited to one user and the way Unraid works is that it allows multiple users to access the container with the UID and GID 99:100 - the UID 99 is not found in the container and it is passed false back to Unraid itself so you can't edit it.
Is it possible to add something like in my startup script that the uid and gid is changed?

https://github.com/ich777/docker-minecraft-basic-server/blob/113d101ccb4d99a91c2e30871bf70a729e0c96e5/scripts/start.sh#L3

Regards, Christoph
P.S: Danke für die super Anwendung! ;)

**Comment by lastzero** [comment]
Thank you! Wouldn't this script fail when the container is **not** started as root? Maybe better to build a special image for this use case?

A workaround, using our existing image, seems possible by mounting such a script and changing the container command name.

**Comment by ich777** [comment]
@lastzero Yes, this would fail if you don't start this as root but as you can see it later stats as the intended user with his UID and GID so you can circumvent this (I start the script of the actual application here:
https://github.com/ich777/docker-minecraft-basic-server/blob/113d101ccb4d99a91c2e30871bf70a729e0c96e5/scripts/start.sh#L30

Yes, that's also possible but that's not really user friendly since Unraid has very much unexperienced users (and also I try to make my containers as userfriendly as possible) that this would be too much to do for them...

I also have to say that your base container also starts as root if I don't pass through any UID & GID.
![grafik](https://user-images.githubusercontent.com/28066518/88200664-bd10e700-cc46-11ea-9ec5-8f3820dbaea5.png)

**Comment by lastzero** [comment]
You can start it as whatever user you want. It's just a binary with some asset files bundled in an archive. That's the point about containers, it's not a virtual machine where you need to create users and configure a complete operating system.

We'll come up with a solution, going to take a couple of days off now after working so much.

**Comment by ich777** [comment]
@lastzero but wouldn't it be better to run it out of the box as non root user? Just a suggestion.
Have you plans to make releases here on github of the binary and the assets that would make it easier for me to put together a container that checks at startup if a newer release is available.

Completely understandable. ;)
Wish you a nice few days off. :)

**Comment by lastzero** [comment]
> @lastzero but wouldn't it be better to run it out of the box as non root user? Just a suggestion.

No, because then it would be a specific user that might not work for everyone (unless you use a dynamic script that again requires root privileges, delays startup and adds attack vectors like a virus scanner that opens every file an attacker provides to check if it is safe).

> Have you plans to make releases here on github of the binary and the assets that would make it easier for me to put together a container that checks at startup if a newer release is available.

As images are basically just archives (like a tar with some metadata), you can do multi-stage builds with Docker which use existing images as template for a new image. We already do this by using the development image as build environment and then copying the binary and assets to the production image. The development image in turn extends an existing Linux distribution. You can extend any of our images in the same way.

Example: https://github.com/photoprism/photoprism/blob/develop/Dockerfile

**Comment by ich777** [comment]
Thanks for the response I will leave it as it is, since it's not what I try to accomplish.

**Comment by lastzero** [comment]
It would in fact be interesting if Unraid and other NAS vendors support Dockerfiles for runnung containers, instead of only Docker Hub image names plus some config options if you're lucky. That way you could easily customize any image by adding files, changing users etc.

We can still provide an image that comes with an init script (as you recommended) if that is what Unraid users expect and are used to. Just trying to explain the technical background and why this is not already the default (as it adds code / logic to solve a problem that is already solved by Docker either by using docker-compose, docker CLI parameters or Dockerfiles).

**Comment by lastzero** [comment]
See also:
- https://medium.com/redbubble/running-a-docker-container-as-a-non-root-user-7d2e00f8ee15
- https://github.com/photoprism/photoprism/blob/develop/docker/demo/Dockerfile

**Comment by ich777** [comment]
> It would in fact be interesting if Unraid and other NAS vendors support Dockerfiles for runnung containers, instead of only Docker Hub image names plus some config options if you're lucky. That way you could easily customize any image by adding files, changing users etc.

Yes they mostly rely on specific templates/configurations also on Unraid DockerCompose is not available.
That's why I'm asking for a init script.



> We can still provide an image that comes with an init script (as you recommended) if that is what Unraid users expect and are used to. Just trying to explain the technical background and why this is not already the default (as it adds code / logic to solve a problem that is already solved by Docker either by using docker-compose, docker CLI parameters or Dockerfiles).

Yes I know the background and work a few years with Docker now... But in Unraid that's a little bit different (also on Synology as far as i know).

**Comment by ich777** [comment]
> Maybe it is as simple as adding `--user ${UID}` to `ExtraParameters`?

Trust me I've tried this, this doesn't work, even if you do --user 99:100 (in this case UID=99 and GID=100)

**Comment by lastzero** [comment]
> Yes they mostly rely on specific templates/configurations also on Unraid DockerCompose is not available.
> That's why I'm asking for a init script.

As mentioned already, a startup script is very easy to add by extending our existing image and then pushing the result to Docker Hub or maybe even an Unraid specific registry.

When you said that's not what you're trying to accomplish, I was trying to come up with a deeper understanding and other potential solutions. I'm not used to implement a solution until I know the actual problem... maybe we need to set up our own test environments after the break.

Anyhow, thanks for your support! :)

**Comment by bobobo1618** [comment]
Has anyone tried raising this with the Unraid folks? Sounds like a limitation of Unraid to me. I don't think the solution to an Unraid limitation is for every Docker image to add a workaround for it.

**Comment by ich777** [comment]
> Has anyone tried raising this with the Unraid folks? Sounds like a limitation of Unraid to me. I don't think the solution to an Unraid limitation is for every Docker image to add a workaround for it.

That's actually a limitation of most NAS Systems since they only provide the dockerd service and minimal capabilities, but don't get me wrong, Unraid is capable of narly allmost containers out there (the implementation in the WebGUI and keeping it simple is the main problem I think).


> As mentioned already, a startup script is very easy to add by extending our existing image and then pushing the result to Docker Hub or maybe even an Unraid specific registry.

I will do a container with an init script, that's not the problem, my main concern was how to figure out if there is a newer version of PhotoPrism available to update the image or the files in the container (normaly I make a container that updates the files so that the container don't have to be updated that frequently - I know that's a little bit uncommon but that's the way I do it... :D ).


> When you said that's not what you're trying to accomplish, I was trying to come up with a deeper understanding and other potential solutions. I'm not used to implement a solution until I know the actual problem... maybe we need to set up our own test environments after the break.

I can also give you access to my Unraid DEV server if you want. ;)
(even if you break it it's not a problem, unraid is setup in a few minutes again)

**Comment by bobobo1618** [comment]
> they [...] provide the dockerd service

If it's actually the `dockerd` service they use, can you "just" manage your containers through the `docker` CLI tool or another web UI? "Incompatible with the Unraid Web UI" is a little better than "Incompatible with Unraid" :)

**Comment by bobobo1618** [comment]
Bonus: If Unraid uses `dockerd`, it should be possible to use `docker-compose`.

**Comment by ich777** [comment]
> Bonus: If Unraid uses `dockerd`, it should be possible to use `docker-compose`.

Yep but not every home user is used to the command line and they (including me creating the containers) try to keep things easy and simple as possible.

Also the same goes for the Synology NAS Devices and I think many other manifacturers too.


> If it's actually the `dockerd` service they use, can you "just" manage your containers through the `docker` CLI tool or another web UI?

docker-compose is not included in the build of Unraid and since all changes are wiped when you restart the server that's not an easy task to include (I also creadted a container to build your own Kernel/Images for Unraid so that you make changes how you like it but that's not the way Unraid is intended to be used).

Yes, everything is possible from the command line but as I said, not every End-User is used to the command line.
They did a very clever job with their templates but there are some limitations...

I will also create a container for PhotoPrism but first I need to know how to get new versions. :)

**Comment by lastzero** [comment]
We can add this to our CI pipeline so that there are official NAS images just like for the Pi.

Not sure if Docker Hub sends notifications for new versions, I think watchtower is polling ever 5 minutes or so.

A major reason to set the UID via Docker is trust, especially when you do it for security reasons and not just convenience like in this case. Otherwise, admins need to trust every image to really give up root and change the process UID as expected.

**Comment by ich777** [comment]
> We can add this to our CI pipeline so that there are official NAS images just like for the Pi.

That would be nice, but I also can create such an image, the only question is how would I know that there is a newer version of PhotoPrism available...


> Not sure if Docker Hub sends notifications for new versions, I think watchtower is polling ever 5 minutes or so.

If Unraid is configured like it's intended then Unraid itself checks for updates. ;)


> A major reason to set the UID via Docker is trust, especially when you do it for security reasons and not just convenience like in this case. Otherwise, admins need to trust every image to really give up root and change the process UID as expected.

I know, but many images are built specificly for NAS systems (I also learned this the hard way...).
Nowerdays much things are customized for End-Users and that's where the problems begin... :D

**Comment by lastzero** [comment]
We'll add it when there is a tested Dockerfile and init script. It's a waste of resources to build it somewhere else as you would have to download the complete image just to add a few bytes (and also continously watch for changes).

**Comment by ich777** [comment]
I tested this also on a Synology NAS with the `--user` parameter and it also isn't happy with that, simply haven't got the rights to delete created files/folders.

I thought again about this and think there was a misunderstanding here are the steps that I would make.
1. In the Dockerfile create a user named whatever (in this example i take `photoprism`)
2. Then start the init script as root, change the UID and GID of the user `photoprism` to the variable `$UID` and `$GID`
3. After that start PhotoPrism as the user `photoprism`

Whish you and your team a nice weekend ;)

**Comment by graciousgrey** [comment]
Can we close this? Or are there some more open to dos?

**Comment by ich777** [comment]
Can be closed from my side.
